README.md
# IDE-PHP package
[![macOS Build Status](https://travis-ci.org/atom/ide-php.svg?branch=master)](https://travis-ci.org/atom/ide-php) [![Windows Build Status](https://ci.appveyor.com/api/projects/status/99y003ssr76ovd17?svg=true)](https://ci.appveyor.com/project/Atom/ide-php/branch/master) [![Dependency Status](https://david-dm.org/atom/ide-php.svg)](https://david-dm.org/atom/ide-php)

PHP language support for Atom-IDE, powered by [FelixFBeckers PHP Language Server](https://github.com/felixfbecker/php-language-server).

Requires [Atom 1.21](https://atom.io/) and a PHP 7.0 runtime installed on your system.

![Screen shot of IDE-PHP](https://user-images.githubusercontent.com/118951/30307874-5a0b03d6-9736-11e7-84d1-55eafe784cda.png)

## Early access

This package is currently an early access release.  You should also install the [atom-ide-ui](https://atom.io/packages/atom-ide-ui) package to expose the functionality within Atom.

## Features

* Auto completion
* Diagnostics (errors & warnings)
* Document outline
* Find references
* Go to definition
* Hover
* Signature help

## Contributing
Always feel free to help out!  Whether it's [filing bugs and feature requests](https://github.com/atom/languageserver-php/issues/new) or working on some of the [open issues](https://github.com/atom/languageserver-php/issues), Atom's [contributing guide](https://github.com/atom/atom/blob/master/CONTRIBUTING.md) will help get you started while the [guide for contributing to packages](https://github.com/atom/atom/blob/master/docs/contributing-to-packages.md) has some extra information.

## License
MIT License.  See [the license](LICENSE.md) for more details.


CHANGELOG.md
## v0.7.18

- Replace line-breaks in HTML description
- Update atom-languageclient to 0.9.9

## v0.7.17

- Update language server to 5.4.6

## v0.7.16

- Update language server to 5.4.4

## v0.7.15

- Update language server to 5.4.2

## v0.7.14

- Bump autocomplete provider model to 4.0.0 so `Class1::` etc. works

## v0.7.13

- Update atom-languageclient to 0.9.4 to allow autocomplete on `->` etc. thanks @pelallemant

## v0.7.12

- Update atom-languageclient to 0.9.3

## v0.7.11

- Language server updated to 5.4.0 which add pseudo-keywords like int, bool, strict_types to completion

## v0.7.10

- Added option to change AutoComplete suggestions pritority (thanks @martinfojtik)

## v0.7.9

- Update atom-languageclient to 0.9.2 which includes better autocomplete

## v0.7.8

- Language server updated to 5.3.7
- Now capable of running on PHP 7.0 again

## v0.7.7

- Now checks for minimum version of PHP 7.1 which is required by language server

## v0.7.6

- Language server updated to 5.3.6 which includes completion and DefinitionResolver fixes (thanks @vinkla)

## v0.7.5

- Language server updated to 5.2.0 which includes Signature Help and fixes (thanks @vinkla)
- Update atom-languageclient to 0.8.3 which includes autocomplete and other fixes

## v0.7.4

- Language server updated to 5.2.0 which includes Signature Help and fixes (thanks @vinkla)

## v0.7.3

- Repackage 5.1.0 language server without prefix (script issue on macOS)

## v0.7.2

- Fix download URL for 5.1.0 language server

## v0.7.1

- Update atom-languageclient to 0.7.0 which includes
  - Sorting autocomplete results
  - Snippet completion type items
  - Busy signals for startup and shutdown
- Fixes broken signals in prior v0.7.0 release

## v0.7.0

- Update PHP language server to 5.1.0

## v0.6.10

- Use busy signal instead of status bar for installation progress
- Composer script added to package.json for ease of preparing new builds
- Update README to remove references to beta

## v0.6.9

- Fix PHP version detection on some systems where stdout delayed
- Address possible async issues in install that resulted in console warnings

## v0.6.8

- Setting for PHP path now points to the PHP binary itself not the folder

## v0.6.7

- Update to 4.6.1 of https://github.com/felixfbecker/php-language-server to fix memory leak
- Checks installed PHP runtime version is 7 or later
- Further improvements to error reporting

## v0.6.6

- Do not display an error dialog when shutting down the language server

## v0.6.5

- Default memory limit to 2GB to avoid issues on 32-bit systems

## v0.6.4

- Allow setting of the memory limit
- Improve error reporting
- Fix description of connection type setting

## v0.6.3

- Allow connection type to be configured in settings
- Deeper error logging via atom-languageclient - enable with atom.config.set('core.debugLSP', true)

## v0.6.2

- Add memory limit setting to work around language server issue on 32-bit systems (set it to 2G)
- Repackage language server to avoid issues with symlinks


lib/main.js
const cp = require('child_process')
const fs = require('fs')
const net = require('net')
const path = require('path')
const {shell} = require('electron')
const {AutoLanguageClient, DownloadFile} = require('atom-languageclient')

const serverDownloadUrl = 'https://github.com/atom/ide-php/files/2580711/php-ls-5.4.6.tar.gz'
const serverDownloadSize = 3056441
const serverLauncher = path.join('felixfbecker', 'language-server', 'bin', 'php-language-server.php')
const minPHPRuntime = '7.0'
const bytesToMegabytes = 1024 * 1024

class PHPLanguageClient extends AutoLanguageClient {
  getGrammarScopes () { return [ 'text.html.php' ] }
  getLanguageName () { return 'PHP' }
  getServerName () { return 'FelixFBecker' }

  provideAutocomplete() {
    const provide = super.provideAutocomplete()
    provide.inclusionPriority = atom.config.get('ide-php.autocompletePriority') === 'lower' ? 2 : 1
    provide.suggestionPriority = atom.config.get('ide-php.autocompletePriority') === 'lower' ? 1 : 2
    return provide
  }
  
  onDidConvertAutocomplete(completionItem, suggestion, request) {
    if (/<[a-z][\s\S]*>/i.test(suggestion.description)) {
      suggestion.descriptionMarkdown = suggestion.description.replace(/\n/g, '');
    }
  }

  startServerProcess () {
    const serverHome = path.join(__dirname, '..', 'vendor')
    return this.getOrInstallServer(serverHome).then(() =>
      this.getConnectionType() === 'socket' ? this.spawnServerWithSocket() : this.spawnServer()
    )
  }

  spawnServerWithSocket () {
    return new Promise((resolve, reject) => {
      let childProcess
      const server = net.createServer(socket => {
        // When the language server connects, grab socket, stop listening and resolve
        this.socket = socket
        server.close()
        resolve(childProcess)
      })
      server.listen(0, '127.0.0.1', () => {
        // Once we have a port assigned spawn the Language Server with the port
        childProcess = this.spawnServer([`--tcp=127.0.0.1:${server.address().port}`])
      })
    })
  }

  checkPHPVersion (command) {
    return new Promise((resolve, reject) => {
      const childProcess = cp.spawn(command, [ '-r', 'echo phpversion();' ])
      var alreadyShownError = false
      childProcess.on('error', err => {
        alreadyShownError = true
        let description = err.code == 'ENOENT'
          ? `No PHP interpreter found at <b>${command}</b>.`
          : `Could not spawn the PHP interpreter <b>${command}</b>.`
        atom.notifications.addError('IDE-PHP could not launch your PHP runtime.', {
          dismissable: true,
          buttons: [
            { text: 'Set PHP Path', onDidClick: () => this.openPackageSettings() },
            { text: 'Download PHP', onDidClick: () => this.downloadPHP() },
          ],
          description: `${description}<p>If you have PHP ${minPHPRuntime} installed please Set PHP Path correctly. If you do not please Download PHP ${minPHPRuntime} or later and install it.</p><code>${err.message}</code>`
        })
        reject()
      })
      let stdErr = '', stdOut = ''
      childProcess.stderr.on('data', chunk => stdErr += chunk.toString())
      childProcess.stdout.on('data', chunk => stdOut += chunk.toString())
      childProcess.on('close', exitCode => {
        if (alreadyShownError) return
        if (exitCode === 0 && stdOut != '') {
          this.logger.debug(`Using PHP ${stdOut} from ${command}`)
          if (this.meetsRequiredVersion(stdOut, minPHPRuntime)) {
            resolve()
          } else {
            atom.notifications.addError(`IDE-PHP requires PHP ${minPHPRuntime} or later but found ${stdOut}`, {
              dismissable: true,
              buttons: [
                { text: 'Set PHP Path', onDidClick: () => this.openPackageSettings() },
                { text: 'Download PHP', onDidClick: () => this.downloadPHP() },
              ],
              description: `If you have PHP ${minPHPRuntime} installed please Set PHP Path correctly. If you do not please Download PHP ${minPHPRuntime} or later and install it.`
            })
            reject()
          }
        } else {
          atom.notifications.addError('IDE-PHP encounted an error using the PHP runtime.', {
            dismissable: true,
            description: stdErr != '' ? `<code>${stdErr}</code>` : `Exit code ${exitCode}`
          })
          reject()
        }
      })
    })
  }

  meetsRequiredVersion (versionString, requiredString) {
    var version = versionString.split('.')
    var required = requiredString.split('.')

    for (var i = 0; i < required.length; i++) {
      if (i >= version.length) return false

      var versionPoint = parseInt(version[i])
      var requiredPoint = parseInt(required[i])

      if (versionPoint < requiredPoint) return false;
      if (versionPoint > requiredPoint) return true;
      // Otherwise check next part via loop
    }

    return true
  }

  downloadPHP () {
    shell.openExternal('https://secure.php.net/downloads.php')
  }

  spawnServer (extraArgs) {
    const command = this.getPHPCommand()

    return this.checkPHPVersion(command).then(() => {
      const serverHome = path.join(__dirname, '..', 'vendor')

      // Temporarily set the default to 2GB as language server defaults to 4GB which fails on 32-bit systems
      const memoryLimit = this.getMemoryLimit() === '-1' ? '2G' : this.getMemoryLimit()
      const args = [ serverLauncher, '--memory-limit=' + memoryLimit ]
      if (extraArgs) {
        args.push(extraArgs)
      }

      this.logger.debug(`starting "${command} ${args.join(' ')}"`)
      const childProcess = cp.spawn(command, args, { cwd: serverHome })
      this.captureServerErrors(childProcess)
      childProcess.on('exit', exitCode => {
        if (exitCode != 0 && exitCode != null) {
          atom.notifications.addError('IDE-PHP language server stopped unexpectedly.', {
            dismissable: true,
            description: this.processStdErr != null ? `<code>${this.processStdErr}</code>` : `Exit code ${exitCode}`
          })
        }
      })
      return childProcess
    })
  }

  openPackageSettings() {
    atom.workspace.open('atom://config/packages/ide-php')
  }

  getOrInstallServer (serverHome) {
    return this.fileExists(path.join(serverHome, serverLauncher))
      .then(doesExist => { if (!doesExist) return this.installServer(serverHome) })
  }

  installServer (serverHome) {
    const localFileName = path.join(serverHome, 'download.tar.gz')
    const decompress = require('decompress')
    const installSignal = this.busySignalService && this.busySignalService.reportBusy('installing...', { revealTooltip: true })
    return this.fileExists(serverHome)
      .then(doesExist => { if (!doesExist) fs.mkdirSync(serverHome) })
      .then(() => DownloadFile(serverDownloadUrl, localFileName, (bytesDone, percent) => installSignal && installSignal.setTitle(`downloading ${Math.floor(serverDownloadSize / bytesToMegabytes)} MB (${percent}% done)`), serverDownloadSize))
      .then(() => installSignal && installSignal.setTitle('unpacking...'))
      .then(() => decompress(localFileName, serverHome))
      .then(() => installSignal && installSignal.dispose())
      .then(() => this.fileExists(path.join(serverHome, serverLauncher)))
      .then(doesExist => { if (!doesExist) throw Error(`Failed to install the ${this.getServerName()} language server`) })
      .then(() => fs.unlinkSync(localFileName))
      .catch(() => new Error(`Failed to install the ${this.getServerName()} language server`))

  }

  getPHPCommand () {
    const configuredPath = atom.config.get('ide-php.phpPath')
    if (configuredPath != null && configuredPath != '') {
        return configuredPath
    }

    const phpHome = process.env['PHP_HOME']
    if (phpHome != null && phpHome != '') {
      return path.join(phpHome, 'php')
    }

    return 'php'
  }

  getMemoryLimit () {
    const memoryLimit = atom.config.get('ide-php.memoryLimit') || '-1'

    if (memoryLimit != '-1' && !/^\d+[KMG]?$/.test(memoryLimit)) {
      atom.notifications.addError(`Invalid memory limit setting of ${memoryLimit}`, {
        dismissable: true,
        buttons: [ { text: 'Set Memory Limit', onDidClick: () => this.openPackageSettings() } ],
        description: 'The memory limit setting should be set to -1, a numeric value or PHP shorthand notation such as 100M, 2G.'
      })
      return '-1'
    }

    return memoryLimit;
  }

  getConnectionType() {
    const configConnectionType = atom.config.get('ide-php.connectionType')
    switch (configConnectionType) {
      case 'auto':    return process.platform === 'win32' ? 'socket' : 'stdio';
      case 'socket':  return 'socket';
      case 'stdio':   return 'stdio';
      default:
        atom.notifications.addWarning('Invalid connection type setting', {
          dismissable: true,
          buttons: [ { text: 'Set Connection Type', onDidClick: () => this.openPackageSettings() } ],
          description: 'The connection type setting should be set to "auto", "socket" or "stdio". "auto" is "socket" on Windows and "stdio" on other platforms.'
        })
    }
  }

  fileExists (path) {
    return new Promise((resolve, reject) => {
      fs.access(path, fs.R_OK, error => {
        resolve(!error || error.code !== 'ENOENT')
      })
    })
  }

  deleteFileIfExists (path) {
    return new Promise((resolve, reject) => {
      fs.unlink(path, error => {
        if (error && error.code !== 'ENOENT') { reject(error) } else { resolve() }
      })
    })
  }
}

module.exports = new PHPLanguageClient()
